src/Utilities/functions.jl

```diff
@@ -135,45 +165,80 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
 The canonical representation of `ScalarAffineFunction([y, x, z, x, z], [2, 1, 3, -2, -3], 5)` is `ScalarAffineFunction([x, y], [-1, 2], 5)`.
 
 """
-function canonical(f::SAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> t.variable_index.value)
-    terms = MOI.ScalarAffineTerm{T}[]
-    for t in sorted_terms
-        if !isempty(terms) && t.variable_index == last(terms).variable_index
-            terms[end] = MOI.ScalarAffineTerm(terms[end].coefficient + t.coefficient, t.variable_index)
-        elseif !iszero(t.coefficient)
-            if !isempty(terms) && iszero(last(terms).coefficient)
-                terms[end] = t
-            else
-                push!(terms, t)
-            end
-        end
+function canonical(f::MOI.ScalarAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.ScalarAffineFunction(copy(f.terms), f.constant))
     end
-    if !isempty(terms) && iszero(last(terms).coefficient)
-        pop!(terms)
+end
+
+function canonical(f::MOI.VectorAffineFunction)
+    if iscanonical(f)
+        f
+    else
+        canonical!(MOI.VectorAffineFunction(copy(f.terms), f.constants))
     end
-    SAF{T}(terms, f.constant)
 end
-function canonical(f::VAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> (t.output_index, t.scalar_term.variable_index.value))
-    terms = MOI.VectorAffineTerm{T}[]
-    for t in sorted_terms
-        if !isempty(terms) && t.output_index == last(terms).output_index && t.scalar_term.variable_index == last(terms).scalar_term.variable_index
-            terms[end] = MOI.VectorAffineTerm(t.output_index, MOI.ScalarAffineTerm(terms[end].scalar_term.coefficient + t.scalar_term.coefficient, t.scalar_term.variable_index))
-        elseif !iszero(t.scalar_term.coefficient)
-            if !isempty(terms) && iszero(last(terms).scalar_term.coefficient)
-                terms[end] = t
+
+
+"""
+    canonical!(f::AbstractFunction)
+
+Convert a function to canonical form in-place, without allocating a copy to hold the result.
+See [`canonical`](@ref).
+"""
+function canonical!(f::MOI.ScalarAffineFunction)
+    merge_duplicates!(f.terms,
+                      t -> t.variable_index.value,
+                      t -> !iszero(t.coefficient),
+                      (t1, t2) -> MOI.ScalarAffineTerm(t1.coefficient + t2.coefficient, t1.variable_index))
+    f
+end
+
+
+function canonical!(f::MOI.VectorAffineFunction)
+    merge_duplicates!(f.terms,
+                      t -> (t.output_index, t.scalar_term.variable_index.value),
+                      t -> !iszero(t.scalar_term.coefficient),
+                      (t1, t2) -> MOI.VectorAffineTerm(t1.output_index,
+                                                       MOI.ScalarAffineTerm(t1.scalar_term.coefficient + t2.scalar_term.coefficient,
+                                                                            t1.scalar_term.variable_index)))
+    f
+end
+
```

Comment:
Using `unsafe_add`, `coefficient` and `termindices` functions, these 2 `canonical!`s can be combined.

Reply:
Yup, I just saw those from your PR and rewrote the function to use them :)